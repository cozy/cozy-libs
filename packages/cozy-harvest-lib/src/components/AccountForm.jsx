import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'
import Field from 'cozy-ui/react/Field'

import AccountFormError from './AccountFormError'
import {
  getEncryptedFieldName,
  getFieldPlaceholder,
  sanitizeSelectProps
} from '../helpers/fields'
import { KonnectorJobError } from '../helpers/konnectors'
import manifest from 'helpers/manifest'
import OAuthForm from './OAuthForm'

const predefinedLabels = [
  'answer',
  'birthdate',
  'code',
  'date',
  'email',
  'firstname',
  'lastname',
  'login',
  'password',
  'phone'
]

// Out of scope labels already used, should be transferred directly in manifests
// in the future.
const legacyLabels = [
  'branchName' // Used in banking konnectors
]

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

export class AccountField extends PureComponent {
  render() {
    const {
      disabled,
      hasError,
      initialValue,
      label,
      name,
      onKeyUp,
      required,
      role,
      t,
      type
    } = this.props

    // Allow manifest to specify predefined label
    const localeKey =
      predefinedLabels.includes(label) || legacyLabels.includes(label)
        ? label
        : name

    const isEditable = !(role === 'identifier' && initialValue)

    const fieldProps = {
      ...this.props,
      autoComplete: 'off',
      className: 'u-m-0', // 0 margin
      disabled: disabled || !isEditable,
      error: hasError,
      fullwidth: true,
      label: t(`fields.${localeKey}.label`, {
        _: t(`legacy.fields.${localeKey}.label`, { _: name })
      }),
      onKeyUp: onKeyUp,
      placeholder: getFieldPlaceholder(
        this.props,
        t(`fields.${name}.placeholder`, { _: '' })
      ),
      side: !required && t('accountForm.fields.optional'),
      size: 'medium'
    }
    const passwordLabels = {
      hideLabel: t('accountForm.password.hide'),
      showLabel: t('accountForm.password.show')
    }
    switch (type) {
      case 'date':
        return (
          <Field
            {...fieldProps}
            placeholder={t(`fields.${name}.placeholder`, {
              _: t(`default.dateFormat`).toLowerCase()
            })}
          />
        )
      case 'dropdown':
        return <Field {...sanitizeSelectProps(fieldProps)} />
      case 'password':
        return (
          <Field
            {...fieldProps}
            // Usting the `new-password` value is the best way to avoid
            // autocomplete for password.
            // See https://stackoverflow.com/a/17721462/1135990
            autoComplete="new-password"
            secondaryLabels={passwordLabels}
          />
        )
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

AccountField.propTypes = {
  hasError: PropTypes.bool,
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  onKeyUp: PropTypes.func,
  role: PropTypes.string,
  type: PropTypes.oneOf(['date', 'dropdown', 'email', 'password', 'text']),
  t: PropTypes.func
}

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const {
      container,
      disabled,
      hasError,
      initialValues,
      manifestFields,
      onKeyUp,
      t
    } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                container={container}
                hasError={hasError}
                disabled={disabled}
                initialValue={
                  initialValues[field.name] ||
                  initialValues[getEncryptedFieldName(field.name)]
                }
                onKeyUp={onKeyUp}
                t={t}
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

AccountFields.propTypes = {
  disabled: PropTypes.bool,
  hasError: PropTypes.bool,
  fillEncrypted: PropTypes.bool,
  manifestFields: PropTypes.object.isRequired,
  onKeyUp: PropTypes.func,
  t: PropTypes.func
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { konnector, lang } = props
    const { locales } = konnector
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const untouched = initialValues && !dirty
    return error || (valid && !untouched)
  }

  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (
      event.code === 'Enter' &&
      this.isSubmittable({ dirty, initialValues, valid })
    ) {
      this.handleSubmit(values, form)
    }
  }

  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const { account, error, konnector, onSubmit, submitting, t } = this.props
    const { fields, oauth } = konnector

    if (oauth) return <OAuthForm initialValues={initialValues} oauth={oauth} />

    const sanitizedFields = manifest.sanitizeFields(fields)
    const defaultValues = manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null

    return (
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            ref={element => {
              container = element
            }}
          >
            {error && (
              <AccountFormError error={error} konnector={konnector} t={t} />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              hasError={
                error &&
                error instanceof KonnectorJobError &&
                error.isLoginError()
              }
              initialValues={initialAndDefaultValues}
              manifestFields={sanitizedFields}
              onKeyUp={event =>
                this.handleKeyUp(event, {
                  dirty,
                  form,
                  initialValues,
                  valid,
                  values
                })
              }
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  konnector: PropTypes.object.isRequired,
  error: PropTypes.object,
  submitting: PropTypes.bool
}

export default translate()(AccountForm)
