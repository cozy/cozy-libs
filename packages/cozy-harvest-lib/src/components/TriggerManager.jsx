import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { withMutations } from 'cozy-client'

import AccountForm from './AccountForm'
import OAuthForm from './OAuthForm'
import TwoFAForm from './TwoFAForm'
import accountsMutations from '../connections/accounts'
import { triggersMutations } from '../connections/triggers'
import filesMutations from '../connections/files'
import permissionsMutations from '../connections/permissions'
import accounts from '../helpers/accounts'
import cron from '../helpers/cron'
import konnectors from '../helpers/konnectors'
import triggers from '../helpers/triggers'
import withLocales from './hoc/withLocales'

const ERRORED = 'ERRORED'
const IDLE = 'IDLE'
const RUNNING = 'RUNNING'
const RUNNING_TWOFA = 'RUNNING_TWOFA'

const MODAL_PLACE_ID = 'coz-harvest-modal-place'

/**
 * Manage a trigger for a given konnector, from account edition to trigger
 * creation and launch.
 * @type {Component}
 */
export class TriggerManager extends Component {
  constructor(props) {
    super(props)
    const { account, trigger } = props

    this.handleNewAccount = this.handleNewAccount.bind(this)
    this.handleOAuthAccountId = this.handleOAuthAccountId.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.closeTwoFAModal = this.closeTwoFAModal.bind(this)
    this.disableSuccessTimer = this.disableSuccessTimer.bind(this)
    this.handleLoginSuccessState = this.handleLoginSuccessState.bind(this)
    this.handleSuccess = this.handleSuccess.bind(this)
    this.handleError = this.handleError.bind(this)
    this.handleTwoFACodeAsked = this.handleTwoFACodeAsked.bind(this)
    this.handleSubmitTwoFACode = this.handleSubmitTwoFACode.bind(this)

    this.jobWatcher = null
    this.accountWatcher = null

    this.state = {
      account,
      error: triggers.getError(trigger),
      status: IDLE,
      trigger
    }
  }

  componentWillUnmount() {
    if (this.jobWatcher) this.jobWatcher.unsubscribeAll()
    if (this.accountWatcher) this.accountWatcher.unsubscribeAll()
  }

  closeTwoFAModal() {
    this.setState({
      status: RUNNING
    })
  }

  disableSuccessTimer() {
    if (this.jobWatcher) this.jobWatcher.disableSuccessTimer()
  }

  handleLoginSuccessState() {
    if (this.jobWatcher) this.jobWatcher.handleSuccess()
  }

  /**
   * Ensure that a trigger will exist, with valid destination folder with
   * permissions and references
   * @return {Object} Trigger document
   */
  async ensureTrigger() {
    const {
      addPermission,
      addReferencesTo,
      createDirectoryByPath,
      createTrigger,
      statDirectoryByPath,
      konnector,
      t
    } = this.props

    const { account, trigger } = this.state

    if (trigger) {
      return trigger
    }

    let folder

    if (konnectors.needsFolder(konnector)) {
      // TODO: remove previous legacy key `default.baseDir`
      // https://github.com/cozy/cozy-libs/issues/637
      const path = konnectors.buildFolderPath(konnector, account, {
        administrative: t('folder.administrative'),
        photos: t('folder.photos')
      })

      folder =
        (await statDirectoryByPath(path)) || (await createDirectoryByPath(path))

      await addPermission(konnector, konnectors.buildFolderPermission(folder))
      await addReferencesTo(konnector, [folder])
    }

    return await createTrigger(
      triggers.buildAttributes({
        account,
        cron: cron.fromKonnector(konnector),
        folder,
        konnector
      })
    )
  }

  /**
   * OAuth Form success handler. OAuthForm retrieves an account id created by the
   * cozy stack
   * @param  {string}  accountId
   */
  async handleOAuthAccountId(accountId) {
    const { findAccount } = this.props
    try {
      const oAuthAccount = await findAccount(accountId)
      return await this.handleNewAccount(oAuthAccount)
    } catch (error) {
      this.handleError(error)
    }
  }

  /**
   * Account creation success handler
   * @param  {Object}  account Created io.cozy.accounts document
   * @return {Object}          io.cozy.jobs document, runned with account data
   */
  async handleNewAccount(account) {
    this.setState({ account, error: null, status: RUNNING })
    const trigger = await this.ensureTrigger()
    this.setState({ trigger })
    this.accountWatcher = this.props.watchKonnectorAccount(account, {
      onTwoFACodeAsked: this.handleTwoFACodeAsked,
      onLoginSuccess: this.handleLoginSuccessState,
      onLoginSuccessHandled: this.disableSuccessTimer
    })
    return await this.launch(trigger)
  }

  handleError(error) {
    this.setState({
      error,
      status: ERRORED
    })

    const { onError } = this.props
    if (typeof onError === 'function') onError(error)
  }

  handleTwoFACodeAsked(statusCode) {
    const { status } = this.state
    if (accounts.isTwoFANeeded(status)) return
    // disable successTimeout since asked Two FA code
    this.disableSuccessTimer()
    this.setState({
      status: statusCode
    })
  }

  async handleSubmitTwoFACode(code) {
    const { konnector, saveAccount } = this.props
    const { account } = this.state
    this.setState({
      error: null,
      status: RUNNING_TWOFA
    })
    try {
      await saveAccount(konnector, accounts.updateTwoFaCode(account, code))
      if (this.jobWatcher) this.jobWatcher.enableSuccessTimer(10000)
    } catch (error) {
      return this.handleError(error)
    }
  }

  async handleSubmit(data = {}) {
    const { konnector, saveAccount } = this.props

    const { account } = this.state
    const isUpdate = !!account

    this.setState({
      error: null,
      status: RUNNING
    })

    try {
      const accountToSave = isUpdate
        ? accounts.mergeAuth(
            accounts.setSessionResetIfNecessary(
              accounts.resetState(account),
              data
            ),
            data
          )
        : accounts.build(konnector, data)
      const savedAccount = accounts.mergeAuth(
        await saveAccount(konnector, accountToSave),
        data
      )
      return await this.handleNewAccount(savedAccount)
    } catch (error) {
      return this.handleError(error)
    }
  }

  /**
   * Handle a success, typically job success or login success
   * @param  {Function} successCallback Typically onLoginSuccess or onSuccess
   * @param  {Array} args            Callback arguments
   */
  async handleSuccess(successCallback) {
    const trigger = await this.refetchTrigger()
    this.setState({ status: IDLE, trigger })
    if (typeof successCallback !== 'function') return
    successCallback(triggers)
  }

  /**
   * Launches a trigger
   * @param  {Object}  trigger io.cozy.triggers document
   * @return {Promise}         [description]
   */
  async launch(trigger) {
    const {
      launchTrigger,
      onLaunch,
      onLoginSuccess,
      onSuccess,
      watchKonnectorJob
    } = this.props
    this.jobWatcher = watchKonnectorJob(await launchTrigger(trigger))
    this.jobWatcher.on('error', this.handleError)
    this.jobWatcher.on('loginSuccess', () => this.handleSuccess(onLoginSuccess))
    this.jobWatcher.on('success', () => this.handleSuccess(onSuccess))

    if (typeof onLaunch === 'function') onLaunch(trigger)
  }

  async refetchTrigger() {
    const { fetchTrigger } = this.props
    const { trigger } = this.state
    try {
      return await fetchTrigger(trigger._id)
    } catch (error) {
      this.setState({ error, running: false })
      throw error
    }
  }

  render() {
    const { konnector, running, showError, modalContainerId } = this.props
    const { account, error, status } = this.state
    const submitting = status === RUNNING || running
    const submittingTwoFA = status === RUNNING_TWOFA
    const waitForTwoFACode = accounts.isTwoFANeeded(status)
    const isTwoFARetryCode = accounts.isTwoFARetry(status)
    const display2FA = waitForTwoFACode || submittingTwoFA || isTwoFARetryCode
    const modalInto = modalContainerId || MODAL_PLACE_ID

    const { oauth } = konnector

    if (oauth) {
      return (
        <OAuthForm
          account={account}
          konnector={konnector}
          onSuccess={this.handleOAuthAccountId}
          submitting={submitting || display2FA}
        />
      )
    }

    return (
      <div>
        <div id={modalInto} />
        <AccountForm
          account={account}
          error={error}
          konnector={konnector}
          onSubmit={this.handleSubmit}
          showError={showError}
          submitting={submitting || display2FA}
        />
        {display2FA && (
          <TwoFAForm
            account={account}
            konnector={konnector}
            dismissAction={this.closeTwoFAModal}
            handleSubmitTwoFACode={this.handleSubmitTwoFACode}
            submitting={submittingTwoFA}
            into={modalInto}
            retryAsked={isTwoFARetryCode && !submittingTwoFA}
          />
        )}
      </div>
    )
  }
}

TriggerManager.propTypes = {
  /**
   * Account document. Used to get intial form values.
   * If no account is passed, AccountForm will use empty initial values.
   * @type {Object}
   */
  account: PropTypes.object,
  /**
   * Konnector document. AccountForm will check the `fields` object to compute
   * fields.
   * @type {Object}
   */
  konnector: PropTypes.object.isRequired,
  /**
   * Indicates if the TriggerManager has to show errors. Sometimes errors may be
   * displayed elsewhere. However, a KonnectorJobError corresponding to a login
   * error is always displayed. Transmitted to AccountForm.
   * @type {Boolean}
   */
  showError: PropTypes.bool,
  /**
   * Existing trigger document to manage.
   * @type {Object}
   */
  trigger: PropTypes.object,
  /**
   * Indicates if the given trigger is already running, i.e. if it has been
   * launched and if an associated job with status 'running' exists.
   * @type {[type]}
   */
  running: PropTypes.bool,
  /**
   * Translation function
   */
  t: PropTypes.func,
  //
  // mutations
  //
  /**
   * Permission mutation
   * @type {Function}
   */
  addPermission: PropTypes.func,
  /**
   * File mutation
   * @type {Function}
   */
  addReferencesTo: PropTypes.func,
  /**
   * Trigger mutation
   * @type {Function}
   */
  createTrigger: PropTypes.func.isRequired,
  /**
   * Trigger mutations
   * @type {Function}
   */
  createDirectoryByPath: PropTypes.func,
  /**
   * Account Mutation, used to retrieve OAuth account
   * @type {Function}
   */
  findAccount: PropTypes.func,
  /**
   * Trigger mutation
   * @type {Function}
   */
  launchTrigger: PropTypes.func.isRequired,
  /**
   * Account mutation
   * @type {Func}
   */
  saveAccount: PropTypes.func.isRequired,
  /**
   * Account mutations
   * @type {Function}
   */
  watchKonnectorAccount: PropTypes.func.isRequired,
  /**
   * Trigger mutations
   * @type {Function}
   */
  statDirectoryByPath: PropTypes.func,
  /**
   * Job mutations
   * @type {Function}
   */
  watchKonnectorJob: PropTypes.func.isRequired,
  //
  // Callbacks
  //
  /**
   * Callback invoke when the konnector job fails
   * @type {Function}
   */
  onError: PropTypes.func,
  /**
   * Callback invoked when the trigger is launch
   * @type {Function}
   */
  onLaunch: PropTypes.func,
  /**
   * Callback invoked when the trigger has been launched and the login to the
   * remote service has succeeded.
   * @type {Function}
   */
  onLoginSuccess: PropTypes.func,
  /**
   * Callback invoked when the trigger has been launched and the job ended
   * successfully.
   * @type {Function}
   */
  onSuccess: PropTypes.func
}

export default withLocales(
  withMutations(
    accountsMutations,
    filesMutations,
    permissionsMutations,
    triggersMutations
  )(TriggerManager)
)
