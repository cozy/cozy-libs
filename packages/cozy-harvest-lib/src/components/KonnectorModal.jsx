import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { withMutations } from 'cozy-client'
import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Button from 'cozy-ui/transpiled/react/Button'
import Infos from 'cozy-ui/transpiled/react/Infos'
import Modal, {
  ModalContent,
  ModalHeader
} from 'cozy-ui/transpiled/react/Modal'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import Icon from 'cozy-ui/transpiled/react/Icon'
import { translate } from 'cozy-ui/transpiled/react/I18n'
import get from 'lodash/get'

import accountMutations from '../connections/accounts'
import triggersMutations from '../connections/triggers'
import * as konnectorsModel from '../helpers/konnectors'
import * as triggersModel from '../helpers/triggers'

import TriggerManager from './TriggerManager'
import AccountSelectBox from './AccountSelectBox/AccountSelectBox'
import AccountsList from './AccountsList/AccountsList'
import KonnectorUpdateInfos from './infos/KonnectorUpdateInfos'
import KonnectorAccountTabs from './KonnectorConfiguration/KonnectorAccountTabs'

/**
 * KonnectorModal can be completely standalone and will use it's internal state to switch between views, or it can be controlled by the parents through props (such as accountId) and callbacks (such as createAction and onAccountChange)
 */
export class KonnectorModal extends PureComponent {
  state = {
    account: null,
    fetching: false,
    fetchingAccounts: false,
    addingAccount: false,
    trigger: null,
    error: null,
    accounts: []
  }

  constructor(props) {
    super(props)
    this.fetchIcon = this.fetchIcon.bind(this)
    this.refetchTrigger = this.refetchTrigger.bind(this)
    this.requestAccountChange = this.requestAccountChange.bind(this)
    this.requestAccountCreation = this.requestAccountCreation.bind(this)
    this.endAccountCreation = this.endAccountCreation.bind(this)
  }
  /**
   * TODO We should not fetchAccounts and fetchAccount since we already have the informations
   * in the props. We kept this sytem for compatibility on the existing override.
   * Next tasks: remove these methods and rewrite the override
   */
  async componentDidMount() {
    await this.fetchAccounts()
    const { accounts } = this.state

    if (this.props.accountId) this.loadSelectedAccountId()
    else if (accounts.length === 1)
      this.requestAccountChange(accounts[0].account, accounts[0].trigger)
  }

  componentWillUnmount() {
    const { into } = this.props
    if (!into || into === 'body') return
    // The Modal is never closed after a dismiss on Preact apps, even if it is
    // not rendered anymore. The best hack we found is to explicitly empty the
    // modal portal container.
    setTimeout(() => {
      try {
        const modalRoot = document.querySelector(into)
        modalRoot.innerHTML = ''
        // eslint-disable-next-line no-empty
      } catch (error) {}
    }, 50)
  }

  componentDidUpdate(prevProps) {
    if (this.props.accountId && this.props.accountId !== prevProps.accountId) {
      this.loadSelectedAccountId()
    }
  }

  requestAccountChange(account, trigger) {
    // This component can either defer the account switching to a parent component through the onAccountChange prop, or handle the change itself if the prop is missing
    const { onAccountChange } = this.props
    return onAccountChange
      ? onAccountChange(account)
      : this.fetchAccount(trigger)
  }

  loadSelectedAccountId() {
    const selectedAccountId = this.props.accountId
    const { accounts } = this.state

    const matchingTrigger = get(
      accounts.find(account => account.account._id === selectedAccountId),
      'trigger'
    )

    if (matchingTrigger) this.fetchAccount(matchingTrigger)
  }

  requestAccountCreation() {
    const { createAction } = this.props
    if (createAction) createAction()
    else this.setState({ addingAccount: true })
  }

  async endAccountCreation(trigger) {
    this.setState({ addingAccount: false })
    const account = await this.fetchAccount(trigger)

    if (account) {
      this.setState(prevState => ({
        ...prevState,
        accounts: [
          ...prevState.accounts,
          {
            account,
            trigger
          }
        ]
      }))
    }
  }

  async fetchAccounts() {
    const triggers = this.props.konnector.triggers.data
    const { findAccount } = this.props
    this.setState({ fetchingAccounts: true })
    try {
      const accounts = (await Promise.all(
        triggers.map(async trigger => {
          return {
            account: await findAccount(triggersModel.getAccountId(trigger)),
            trigger
          }
        })
      )).filter(({ account }) => !!account)
      this.setState({ accounts, fetchingAccounts: false, error: null })
    } catch (error) {
      this.setState({ error, fetchingAccounts: false })
    }
  }

  async fetchAccount(trigger) {
    const { findAccount } = this.props
    this.setState({ fetching: true })

    try {
      const account = await findAccount(triggersModel.getAccountId(trigger))
      this.setState({
        account,
        trigger
      })
      return account
    } catch (error) {
      this.setState({
        error
      })
    } finally {
      this.setState({
        fetching: false,
        error: null
      })
    }
  }

  fetchIcon() {
    const { client } = this.context
    const { konnector } = this.props
    return client.stackClient.getIconURL({
      type: 'konnector',
      slug: konnector.slug
    })
  }

  async refetchTrigger() {
    const { fetchTrigger } = this.props
    const { trigger } = this.state

    const upToDateTrigger = await fetchTrigger(trigger._id)
    this.setState({
      trigger: upToDateTrigger
    })
  }

  render() {
    const { dismissAction, konnector, into, t } = this.props
    const { account, accounts, addingAccount } = this.state

    return (
      <Modal
        dismissAction={dismissAction}
        mobileFullscreen
        size="small"
        into={into}
        closable={false}
      >
        <ModalHeader className="u-pr-2">
          <div className="u-flex u-flex-row u-w-100 u-flex-items-center">
            <div className="u-w-3 u-h-3 u-mr-half">
              <AppIcon fetchIcon={this.fetchIcon} />
            </div>
            <div className="u-flex-grow-1 u-mr-half">
              <h3 className="u-title-h3 u-m-0">{konnector.name}</h3>

              {accounts.length > 0 && account && !addingAccount && (
                <AccountSelectBox
                  selectedAccount={account}
                  accountsList={accounts}
                  onChange={option => {
                    this.requestAccountChange(option.account, option.trigger)
                  }}
                  onCreate={this.requestAccountCreation}
                />
              )}
            </div>
            <Button
              icon={<Icon icon={'cross'} size={'24'} />}
              onClick={dismissAction}
              iconOnly
              label={t('close')}
              subtle
              theme={'secondary'}
            />
          </div>
        </ModalHeader>
        <ModalContent>{this.renderModalContent()}</ModalContent>
      </Modal>
    )
  }

  renderModalContent() {
    const { dismissAction, konnector, t } = this.props
    const {
      account,
      accounts,
      error,
      fetching,
      fetchingAccounts,
      trigger,
      addingAccount
    } = this.state

    if (fetching || fetchingAccounts) {
      return (
        <Spinner
          size="xxlarge"
          className="u-flex u-flex-justify-center u-pv-3"
        />
      )
    } else if (error) {
      return (
        <Infos
          actionButton={
            <Button theme="danger">{t('modal.konnector.error.button')}</Button>
          }
          title={t('modal.konnector.error.title')}
          text={t('modal.konnector.error.description', error)}
          isImportant
        />
      )
    } else if (addingAccount) {
      return (
        <div className="u-pt-1-half">
          <TriggerManager
            konnector={konnector}
            onLoginSuccess={this.endAccountCreation}
            onSuccess={this.endAccountCreation}
          />
        </div>
      )
    } else if (!account) {
      return (
        <>
          {konnectorsModel.hasNewVersionAvailable(konnector) && (
            <KonnectorUpdateInfos className="u-mb-1" konnector={konnector} />
          )}
          <AccountsList
            accounts={accounts}
            konnector={konnector}
            onPick={option => {
              this.requestAccountChange(option.account, option.trigger)
            }}
            addAccount={this.requestAccountCreation}
          />
        </>
      )
    } else {
      return (
        <KonnectorAccountTabs
          konnector={konnector}
          trigger={trigger}
          account={account}
          onAccountDeleted={dismissAction}
          addAccount={this.requestAccountCreation}
          refetchTrigger={this.refetchTrigger}
        />
      )
    }
  }
}

KonnectorModal.propTypes = {
  into: PropTypes.string,
  konnector: PropTypes.shape({
    slug: PropTypes.string,
    name: PropTypes.string,
    vendor_link: PropTypes.string,
    triggers: PropTypes.shape({
      data: PropTypes.arrayOf(PropTypes.object)
    })
  }).isRequired,
  findAccount: PropTypes.func.isRequired,
  fetchTrigger: PropTypes.func,
  dismissAction: PropTypes.func.isRequired,
  createAction: PropTypes.func,
  onAccountChange: PropTypes.func,
  t: PropTypes.func.isRequired
}

KonnectorModal.defaultProps = {
  into: 'body'
}

KonnectorModal.contextTypes = {
  client: PropTypes.object.isRequired
}

export default withMutations(accountMutations, triggersMutations)(
  translate()(KonnectorModal)
)
