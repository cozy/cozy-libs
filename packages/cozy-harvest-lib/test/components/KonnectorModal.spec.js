/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'

import { KonnectorModal } from 'components/KonnectorModal'
import TriggerManager from 'components/TriggerManager'

const t = key => key

const findAccountMock = jest.fn().mockImplementation(() => ({
  id: 123,
  doctype: 'io.cozy.accounts'
}))

describe('KonnectorModal', () => {
  const mockKonnector = {
    slug: 'mock',
    name: 'Mock',
    triggers: {
      data: [{ id: 784, doctype: 'io.cozy.triggers' }]
    }
  }
  const props = {
    findAccount: findAccountMock,
    dismissAction: jest.fn(),
    createAction: jest.fn(),
    onSuccess: jest.fn(),
    konnector: mockKonnector,
    t
  }
  const shallowOptions = {
    context: {
      client: {}
    }
  }

  it('should show a spinner while loading', () => {
    const component = shallow(<KonnectorModal {...props} />, shallowOptions)

    const spinner = component
      .dive()
      .find('ModalContent')
      .childAt(0)
      .dive()
    expect(spinner.getElement()).toMatchSnapshot()
  })

  it('should show the content', () => {
    const component = shallow(<KonnectorModal {...props} />, shallowOptions)
    component.setState({ fetching: false })

    const content = component.dive().find('ModalContent')
    expect(content.getElement()).toMatchSnapshot()
  })

  it('should pass trigger if konnector has triggers', async () => {
    const component = await shallow(
      <KonnectorModal {...props} />,
      shallowOptions
    )
    const manager = component.find(TriggerManager)
    expect(manager.props().trigger).toBeDefined()
  })
})
