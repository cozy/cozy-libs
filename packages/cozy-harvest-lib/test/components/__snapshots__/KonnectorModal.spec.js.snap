// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KonnectorModal adding an account should render the form when controlled by state 1`] = `
<ModalContent>
  <div
    className="u-pt-1-half"
  >
    <LegacyTriggerManager
      konnector={
        Object {
          "name": "Mock",
          "slug": "mock",
          "triggers": Object {
            "data": Array [
              Object {
                "_id": 784,
                "doctype": "io.cozy.triggers",
              },
            ],
          },
        }
      }
      onLoginSuccess={[Function]}
      onSuccess={[Function]}
    />
  </div>
</ModalContent>
`;

exports[`KonnectorModal should render the selected account via a prop 1`] = `
<ModalContent>
  <withClient(Wrapper)
    account={
      Object {
        "_id": "123",
        "doctype": "io.cozy.accounts",
      }
    }
    addAccount={[Function]}
    konnector={
      Object {
        "name": "Mock",
        "slug": "mock",
        "triggers": Object {
          "data": Array [
            Object {
              "_id": "784",
              "doctype": "io.cozy.triggers",
            },
            Object {
              "_id": "872",
              "doctype": "io.cozy.triggers",
            },
          ],
        },
      }
    }
    onAccountDeleted={[MockFunction]}
    refetchTrigger={[Function]}
    trigger={
      Object {
        "_id": "784",
        "doctype": "io.cozy.triggers",
      }
    }
  />
</ModalContent>
`;

exports[`KonnectorModal should show a spinner while loading 1`] = `
<Spinner
  className="u-flex u-flex-justify-center u-pv-3"
  color="var(--primaryColor)"
  loadingType=""
  middle={false}
  noMargin={false}
  size="xxlarge"
/>
`;

exports[`KonnectorModal should show an error view 1`] = `
<ModalContent>
  <Infos
    actionButton={
      <DefaultButton
        theme="danger"
      >
        modal.konnector.error.button
      </DefaultButton>
    }
    icon={null}
    isImportant={true}
    text="modal.konnector.error.description"
    title="modal.konnector.error.title"
  />
</ModalContent>
`;

exports[`KonnectorModal should show the configuration view of a single account 1`] = `
<ModalContent>
  <withClient(Wrapper)
    account={
      Object {
        "_id": "123",
        "doctype": "io.cozy.accounts",
      }
    }
    addAccount={[Function]}
    konnector={
      Object {
        "name": "Mock",
        "slug": "mock",
        "triggers": Object {
          "data": Array [
            Object {
              "_id": 784,
              "doctype": "io.cozy.triggers",
            },
          ],
        },
      }
    }
    onAccountDeleted={[MockFunction]}
    refetchTrigger={[Function]}
    trigger={
      Object {
        "_id": 784,
        "doctype": "io.cozy.triggers",
      }
    }
  />
</ModalContent>
`;

exports[`KonnectorModal should show the list of accounts 1`] = `
<ModalContent>
  <React.Fragment>
    <Wrapper
      accounts={
        Array [
          Object {
            "account": Object {
              "_id": "123",
              "doctype": "io.cozy.accounts",
            },
            "trigger": Object {
              "_id": "784",
              "doctype": "io.cozy.triggers",
            },
          },
          Object {
            "account": Object {
              "_id": "123",
              "doctype": "io.cozy.accounts",
            },
            "trigger": Object {
              "_id": "872",
              "doctype": "io.cozy.triggers",
            },
          },
        ]
      }
      addAccount={[Function]}
      konnector={
        Object {
          "name": "Mock",
          "slug": "mock",
          "triggers": Object {
            "data": Array [
              Object {
                "_id": "784",
                "doctype": "io.cozy.triggers",
              },
              Object {
                "_id": "872",
                "doctype": "io.cozy.triggers",
              },
            ],
          },
        }
      }
      onPick={[Function]}
    />
  </React.Fragment>
</ModalContent>
`;

exports[`KonnectorModal switching account should show the add account view when controlled by state 1`] = `
<ModalContent>
  <withClient(Wrapper)
    account={
      Object {
        "_id": "123",
        "doctype": "io.cozy.accounts",
      }
    }
    addAccount={[Function]}
    konnector={
      Object {
        "name": "Mock",
        "slug": "mock",
        "triggers": Object {
          "data": Array [
            Object {
              "_id": 784,
              "doctype": "io.cozy.triggers",
            },
          ],
        },
      }
    }
    onAccountDeleted={[MockFunction]}
    refetchTrigger={[Function]}
    trigger={
      Object {
        "_id": "abc",
      }
    }
  />
</ModalContent>
`;
